geomesa-convert/geomesa-convert-common/src/main/scala/org/locationtech/geomesa/convert2/AbstractConverter.scala

```diff
@@ -0,0 +1,291 @@
+/***********************************************************************
+ * Copyright (c) 2013-2018 Commonwealth Computer Research, Inc.
+ * All rights reserved. This program and the accompanying materials
+ * are made available under the terms of the Apache License, Version 2.0
+ * which accompanies this distribution and is available at
+ * http://www.opensource.org/licenses/apache2.0.php.
+ ***********************************************************************/
+
+package org.locationtech.geomesa.convert2
+
+import java.io.{IOException, InputStream}
+import java.nio.charset.Charset
+
+import com.typesafe.config.Config
+import com.typesafe.scalalogging.LazyLogging
+import org.geotools.factory.Hints
+import org.locationtech.geomesa.convert.ErrorMode.ErrorMode
+import org.locationtech.geomesa.convert.ParseMode.ParseMode
+import org.locationtech.geomesa.convert._
+import org.locationtech.geomesa.convert2.transforms.Expression
+import org.locationtech.geomesa.features.ScalaSimpleFeature
+import org.locationtech.geomesa.utils.collection.CloseableIterator
+import org.opengis.feature.simple.{SimpleFeature, SimpleFeatureType}
+
+import scala.collection.mutable.{ArrayBuffer, ListBuffer}
+import scala.util.control.NonFatal
+
+/**
+  * Abstract converter implementation. Typically paired with an AbstractConverterFactory. If so, needs to have
+  * a default constructor consisting of (targetSft, config, fields, options), so that the AbstractConverterFactory
+  * can instantiate it via reflection.
+  *
+  * Subclasses need to implement `read` to parse the underlying input stream into raw values that will be
+  * transformed to simple features.
+  *
+  * @param targetSft simple feature type
+  * @param config converter config
+  * @param fields converter fields
+  * @param options converter options
+  * @tparam C config binding
+  * @tparam F field binding
+  * @tparam O options binding
+  */
+abstract class AbstractConverter[C <: ConverterConfig, F <: Field, O <: ConverterOptions]
+  (override val targetSft: SimpleFeatureType, val config: C, fields: Seq[F], options: O)
+    extends SimpleFeatureConverter with LazyLogging {
+
+  private val requiredFields: Array[Field] =
+    AbstractConverter.requiredFields(targetSft, fields, config.userData.values.toSeq ++ config.idField.toSeq)
+
+  private val requiredFieldsCount: Int = requiredFields.length
+
+  private val requiredFieldsIndices: Array[Int] = requiredFields.map(f => targetSft.indexOf(f.name))
+
+  /**
+    * Read values for simple features out of the input stream. This should be lazily evaluated,
+    * so that any exceptions occur in the call to `hasNext`. If there is any sense of 'lines',
```

Comment:
Why do we want exceptions on `hasNext`?  Doesn't it make more sense to throw them on `next`, even if the feature extraction is lazy?

Reply:
the main reason is that if there is an error when fetching the next value, then we might hit the end of the iterator. So if we return true from `hasNext`, then catch, handle and swallow an error in `next`, we don't have anything to return and we've broken the iterator contract.